packages/core/src/agent/getOutboundMessageContext.ts

```diff
@@ -0,0 +1,247 @@
+  if (connectionRecord) {
+    agentContext.config.logger.debug(
+      `Creating outbound message context for message ${message.id} with connection ${connectionRecord.id}`
+    )
+    return new OutboundMessageContext(message, {
+      agentContext,
+      associatedRecord,
+      connection: connectionRecord,
+    })
+  }
+
+  if (!lastReceivedMessage) {
+    throw new AriesFrameworkError(
+      'No connection record and no lastReceivedMessage was supplied. For connection-less exchanges the lastReceivedMessage is required.'
+    )
+  }
+
+  if (!associatedRecord) {
+    throw new AriesFrameworkError(
+      'No associated record was supplied. This is required for connection-less exchanges to store the associated ~service decorator on the message.'
+    )
+  }
+
+  // Connectionless
+  return getConnectionlessOutboundMessageContext(agentContext, {
+    message,
+    associatedRecord,
+    lastReceivedMessage,
+    lastSentMessage,
+  })
+}
+
+export async function getConnectionlessOutboundMessageContext(
+  agentContext: AgentContext,
+  {
+    message,
+    lastReceivedMessage,
+    lastSentMessage,
+    associatedRecord,
+  }: {
+    message: AgentMessage
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    associatedRecord: BaseRecord<any, any, any>
+    lastReceivedMessage: AgentMessage
+    lastSentMessage?: AgentMessage
+  }
+) {
+  agentContext.config.logger.debug(
+    `Creating outbound message context for message ${message.id} using connection-less exchange`
+  )
+  let ourService = lastSentMessage?.service?.resolvedDidCommService
+  let recipientService = lastReceivedMessage.service?.resolvedDidCommService
+
+  agentContext.config.logger.debug(
+    `Looking for out-of-band record for connection-less exchange for message ${message.id}`
+  )
+  const outOfBandRepository = agentContext.dependencyManager.resolve(OutOfBandRepository)
+  const outOfBandService = agentContext.dependencyManager.resolve(OutOfBandService)
+
+  const outOfBandRecord = await outOfBandRepository.findSingleByQuery(agentContext, {
+    invitationRequestsThreadIds: [message.threadId],
+  })
+
+  // Check if valid
+  if (outOfBandRecord?.role === OutOfBandRole.Sender) {
+    // Extract ourService from the oob record if not on a previous message
+    if (!ourService) {
+      ourService = await outOfBandService.getResolvedServiceForOutOfBandServices(
+        agentContext,
+        outOfBandRecord.outOfBandInvitation.getServices()
+      )
+    }
+
+    if (!recipientService) {
+      throw new AriesFrameworkError(
+        `Could not find a service to send the message to. Please make sure the connection has a service or provide a service to send the message to.`
+      )
+    }
+
+    // We have created the oob record with a message, that message should be provided here as well
+    if (!lastSentMessage) {
+      throw new AriesFrameworkError('Must have lastSentMessage when out of band record has role Sender')
+    }
```

Comment:
Sections like this would be quite nice to have as a separate function even if we would only use it once.